### PR TITLE
fix(controller): disallow unauthorized users from modifying apps

### LINKED ISSUE #2820: Any user can add a domain to any app
See https://github.com/deis/deis/pull/2818#issuecomment-68319676

```console
$ deis register ...
$ deis create
$ git push deis master
$ deis logout
$ deis register ...
$ deis domains:add test.lol.com
Adding test.lol.com to quinoa-icehouse... done
```

### PR DESCRIPTION
**Pull request by bacongobbler**
Authenticated users who were not given explicit permission
to view or modify an app were still allowed to create new releases,
run commands, retrieve logs etc. This makes those requests return a 404
if the user should not be able to see the application.

Please let me know if I missed any loopholes and I will happily correct them.

closes #2820, refs #2811

### CODE REVIEW

**Comment by carmstrong** [comment]
```
/home/jenkins/workspace/test-integration/src/github.com/deis/deis/tests
apps:run echo Hello, 世界
Running `echo Hello, 世界`...
Hello, 世界
--- FAIL: TestApps (137.27s)
	itutils.go:272: <nil>
FAIL
exit status 1
FAIL	github.com/deis/deis/tests	137.273s
godep: go exit status 1
make[1]: *** [test-full] Error 1
make[1]: Leaving directory `/home/jenkins/workspace/test-integration/src/github.com/deis/deis/tests'
make: *** [test-integration] Error 2
```

Weird, considering #2813 passed Jenkins. Maybe a red herring? I'll run Jenkins again.

**Comment by carmstrong** [comment]
Failed again in exactly the same spot. Did this PR modify anything that would affect that?

**Comment by carmstrong** [comment]
@bacongobbler did this pass integration tests for you locally? That would rule out this PR.

**Comment by carmstrong** [comment]
Code LGTM, although I didn't comb through the code to see if we missed anything. @mboersma does it look like we missed anything, at first glance?

**Comment by mboersma** [comment]
I looked through the models.py and I couldn't think of similar security issues; this seems to cover it.

@bacongobbler pointed out that since this changes 403 to 404 responses, it should probably require an API version bump. Also we may want to reword this to use "unauthorized" instead.

LGTM, pending resolution of those two items.

**Comment by bacongobbler** [comment]
bumped the controller API version to v1.1.2 in response to the changes as well as an added note in the API docs. Renamed all references of "evil" to "unauthorized" or "foo/bar" as well. Please take another look :)

File: docs/reference/api-v1.1.rst [hunk 1]
```diff
@@ -16,6 +16,8 @@ What's New
 
 **New!** All controller responses now return the ``X_DEIS_PLATFORM_VERSION`` header.
 
+**New!** Users see a 404 response when modifying applications they are unauthorized to see.
+
```
Comment (mboersma) [inline docs/reference/api-v1.1.rst]: Should we also update all the example requests in this doc to say `X_DEIS_API_VERSION: 1.1.2`? Looks like we didn't do that for the 1.1.1 API change, but I think that would be consistent. I've been updating the `X_DEIS_PLATFORM_VERSION:` in the examples with each release.

Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: yes, good point.

Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: done!

Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: d'oh, you meant X_DEIS_API_VERSION. In the response, the patch version gets cut off (clients should not care what happens in patch releases as they're just patches), so `1.1` is intended: https://github.com/deis/deis/blob/6f00ed544d8d008ae8f5769b2581e7e1897f9249/controller/api/middleware.py#L32

No change necessary, then :)

Comment (mboersma) [inline docs/reference/api-v1.1.rst]: Oh, ok, that makes sense. Sorry for my confusion. LGTM again.

**Comment by carmstrong** [comment]
:sheep:
